avaje-sigma/src/main/java/io/avaje/sigma/routes/SpiRoutes.java

```diff
@@ -11,10 +13,10 @@ public interface SpiRoutes {
   Entry match(Router.HttpMethod type, String pathInfo);
 
   /** Execute all appropriate before filters for the given request URI. */
-  void before(String pathInfo, HttpContext ctx) throws Exception;
+  void before(String pathInfo, SigmaContext ctx) throws Exception;
```

Comment:
Why use the concrete SigmaContext here in the API? I'm thinking that `handlersSkipped()` probably should be public API of HttpContext.

That is, for me I'm aiming for an API to be interfaces (purely abstract). It looks like HttpContext should still work if `handlersSkipped()` was made part of the API.

Once we use concrete classes like SigmaContext in the API it becomes harder for folks to use test doubles etc and restricts the ability to change the implementation over time so trying to keep the API as interfaces is imo good practice.

Reply:
> Once we use concrete classes like SigmaContext in the API it becomes harder for folks to use test doubles etc and restricts the ability to change the implementation over time so trying to keep the API as interfaces is imo good practice.

Possibly, but SPIRoutes is an internal API not meant to be used externally